services/datamanager/builtin/sync_test.go

```diff
@@ -0,0 +1,642 @@
+			manualSync:           true,
+			scheduleSyncDisabled: false,
+			serviceFail:          false,
+		},
+		{
+			name:                 "If an error response is received from the backend, local files should not be deleted.",
+			manualSync:           false,
+			scheduleSyncDisabled: false,
+			serviceFail:          false,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			// Set up server.
+			tmpDir, err := os.MkdirTemp("", "")
+			test.That(t, err, test.ShouldBeNil)
+			defer func() {
+				err := os.RemoveAll(tmpDir)
+				test.That(t, err, test.ShouldBeNil)
+			}()
+
+			var failFor int
+			if tc.serviceFail {
+				failFor = 100
+			}
+			rpcServer, mockService := buildAndStartLocalSyncServer(t, 0, failFor)
+			defer func() {
+				err := rpcServer.Stop()
+				test.That(t, err, test.ShouldBeNil)
+			}()
+
+			// Set up data manager.
+			dmsvc := newTestDataManager(t)
+			dmsvc.SetSyncerConstructor(getTestSyncerConstructor(rpcServer))
+			cfg := setupConfig(t, enabledTabularCollectorConfigPath)
+
+			// Set up service config.
+			svcConfig, ok, err := getServiceConfig(cfg)
+			test.That(t, err, test.ShouldBeNil)
+			test.That(t, ok, test.ShouldBeTrue)
+			svcConfig.CaptureDisabled = true
+			svcConfig.ScheduledSyncDisabled = tc.scheduleSyncDisabled
+			svcConfig.SyncIntervalMins = 0.001
+			svcConfig.AdditionalSyncPaths = []string{tmpDir}
+
+			// Start dmsvc.
+			dmsvc.SetWaitAfterLastModifiedMillis(testLastModifiedMillis)
+			err = dmsvc.Update(context.Background(), cfg)
+			test.That(t, err, test.ShouldBeNil)
+
+			// Write some files to the path.
+			fileContents := make([]byte, datasync.UploadChunkSize*4)
+			fileContents = append(fileContents, []byte("happy cows come from california")...)
+			tmpFile, err := os.Create(filepath.Join(tmpDir, fileName))
+			test.That(t, err, test.ShouldBeNil)
+			_, err = tmpFile.Write(fileContents)
+			test.That(t, err, test.ShouldBeNil)
+			time.Sleep(time.Millisecond * 100)
+
+			// Call manual sync if desired.
+			if tc.manualSync {
+				err = dmsvc.Sync(context.Background(), nil)
+				test.That(t, err, test.ShouldBeNil)
+			}
+			time.Sleep(syncTime)
+
+			// Validate error and URs.
+			remainingFiles := getAllFilePaths(tmpDir)
+			if tc.serviceFail {
+				// Error case.
+				test.That(t, len(remainingFiles), test.ShouldEqual, 1)
+			} else {
+				// Validate first metadata message.
+				test.That(t, len(mockService.getFileUploadRequests()), test.ShouldBeGreaterThan, 0)
+				actMD := mockService.getFileUploadRequests()[0].GetMetadata()
+				test.That(t, actMD, test.ShouldNotBeNil)
+				test.That(t, actMD.Type, test.ShouldEqual, v1.DataType_DATA_TYPE_FILE)
+				test.That(t, actMD.FileName, test.ShouldEqual, fileName)
+				test.That(t, actMD.FileExtension, test.ShouldEqual, fileExt)
+				test.That(t, actMD.GetPartId(), test.ShouldEqual, cfg.Cloud.ID)
+
+				// Validate ensuing data messages.
```

Comment:
TODO?

Reply:
Ah good catch, totally forgot to add. There now